File: leetcode/heap/1578. Minimum Deletion Cost to Avoid Repeating Letters/py/main.py
class Solution:
    # Time O(N)
    # Space O(1)
    def minCost(self, s, cost):
        if not s: return 0
        n = len(s)
        ans, i = 0, 0

        while i < n:
            # two pointers
            j = i + 1

            while j < n and s[i] == s[j]:
                ans += min(cost[i], cost[j])

                if cost[i] < cost[j]:
                    # swap because we need to remove the char with less cost
                    i = j

                j += 1

                if j < n and s[i] != s[j]:
                    i = j - 1
                    break

            if j == n:
                break

            i += 1

        return ans

File: leetcode/heap/1578. Minimum Deletion Cost to Avoid Repeating Letters/js/main.js
// Time O(NLogN)
// Space O(N)
const minCost = (s, cost) => {
  let n = s.length;
  let pq = new PriorityQueue({ comparator: (a, b) => a - b });
  let res = 0;
  pq.offer(cost[0]);

  for (let i = 1; i < n; i++) {
    if (s[i - 1] == s[i]) {
      pq.offer(cost[i]);
    } else {
      if (pq.size() > 1) {
        while (pq.size() > 1) {
          res += pq.poll();
        }
      }

      pq.poll();
      pq.offer(cost[i]);
    }
  }

  while (pq.size() > 1) {
    res += pq.poll();
  }

  return res;
};
